File: lab_6/f.py
import sys

grade_to_gpa = {
    "A+": 4.0, "A": 3.75,
    "B+": 3.5, "B": 3.0,
    "C+": 2.5, "C": 2.0,
    "D+": 1.5, "D": 1.0,
    "F": 0.0
}

input = sys.stdin.readline
n = int(input())
students = []

for _ in range(n):
    data = input().split()
    lastname = data[0]
    firstname = data[1]
    k = int(data[2])

    grades = data[3::2]
    credits = list(map(int, data[4::2]))

    gpas = [grade_to_gpa[g] for g in grades]
    gpa = sum(g * c for g, c in zip(gpas, credits)) / sum(credits)

    students.append((gpa, lastname, firstname))

def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = []
    middle = []
    right = []
    for x in arr:
        if x < pivot:
            left.append(x)
        elif x > pivot:
            right.append(x)
        else:
            middle.append(x)
    return quick_sort(left) + middle + quick_sort(right)


students = quick_sort(students)

for gpa, lastname, firstname in students:
    sys.stdout.write(f"{lastname} {firstname} {gpa:.3f}\n")

File: lab_6/f.cpp
#include <iostream>
#include <string>
using namespace std;

struct Student {
    string last;
    string first;
    double gpa;
};

double gradeToGpa(string g) {
    if (g == "A+") return 4.0;
    if (g == "A") return 3.75;
    if (g == "B+") return 3.5;
    if (g == "B") return 3.0;
    if (g == "C+") return 2.5;
    if (g == "C") return 2.0;
    if (g == "D+") return 1.5;
    if (g == "D") return 1.0;
    if (g == "F") return 0.0;
    return 0.0;
}

// Қолмен quick sort
void quickSort(Student arr[], int left, int right) {
    int i = left, j = right;
    Student pivot = arr[(left + right) / 2];
    while (i <= j) {
        while (
            (arr[i].gpa < pivot.gpa) ||
            (arr[i].gpa == pivot.gpa && arr[i].last < pivot.last) ||
            (arr[i].gpa == pivot.gpa && arr[i].last == pivot.last && arr[i].first < pivot.first)
        ) i++;
        while (
            (arr[j].gpa > pivot.gpa) ||
            (arr[j].gpa == pivot.gpa && arr[j].last > pivot.last) ||
            (arr[j].gpa == pivot.gpa && arr[j].last == pivot.last && arr[j].first > pivot.first)
        ) j--;
        if (i <= j) {
            Student tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
            i++; j--;
        }
    }
    if (left < j) quickSort(arr, left, j);
    if (i < right) quickSort(arr, i, right);
}

void printGpa(double g) {
    int whole = (int)g;
    int frac = (int)((g - whole) * 1000 + 0.5);
    if (frac == 1000) { whole++; frac = 0; }
    cout << whole << ".";
    if (frac < 10) cout << "00";
    else if (frac < 100) cout << "0";
    cout << frac;
}

int main() {
    int n;
    cin >> n;
    Student s[100005];

    for (int i = 0; i < n; i++) {
        string last, first;
        int m;
        cin >> last >> first >> m;
        double sum = 0, credits = 0;

        for (int j = 0; j < m; j++) {
            string grade;
            int c;
            cin >> grade >> c;
            sum += gradeToGpa(grade) * c;
            credits += c;
        }

        s[i].last = last;
        s[i].first = first;
        s[i].gpa = sum / credits;
    }

    quickSort(s, 0, n - 1);

    for (int i = 0; i < n; i++) {
        cout << s[i].last << " " << s[i].first << " ";
        printGpa(s[i].gpa);
        cout << "\n";
    }

    return 0;
}